### PR TITLE
feat: support fused selected-token logprobs for AutoModel

### PR DESCRIPTION
**Pull request by mohamed-alansary**
# What does this PR do ?

Adds an opt-in AutoModel path for computing selected next-token logprobs directly from hidden states and the LM-head weight, avoiding full `[B, S, V]` logits materialization for logprob/loss postprocessing.

# Issues

N/A

# Usage

For AutoModel policies, enable the feature with top-level policy keys:

```yaml
policy:
  dtensor_cfg:
    enabled: true
  use_fused_linear_logprobs: true
  fused_linear_logprobs_chunk_size: 256
```

This is intended for SFT/DPO/GRPO-style logprob losses where only selected-token logprobs are needed.

# Details

This PR extends the existing fused selected-token logprob optimization to the AutoModel backend.

Main changes:

- Adds AutoModel policy config keys:
  - `policy.use_fused_linear_logprobs`
  - `policy.fused_linear_logprobs_chunk_size`
- Validates unsupported combinations early:
  - context parallelism
  - sequence packing
  - top-k/top-p training-time filtering
  - missing fused logprob chunk size
- Finds common LM-head locations:
  - `lm_head`
  - `language_model.lm_head`
  - `model.lm_head`
- Captures hidden states entering the LM head with a forward pre-hook.
- Computes selected next-token logprobs with `from_parallel_hidden_states_to_logprobs`.
- Supports both:
  - `LogprobsPostProcessor`
  - `LossPostProcessor` when the loss input type is `LossInputType.LOGPROB`
- Keeps full-logits paths unchanged for reward scores, top-k logits, and unsupported sampling filters.
- Updates SFT, DPO, and GRPO docs with AutoModel usage and limitations.
- Adds focused unit coverage for setup, validation, precomputed selected-logprob handling, and fused forward wiring.

# Design decision: `lm_head.forward` wrapper

This PR installs an opt-in wrapper around `lm_head.forward` when `policy.use_fused_linear_logprobs` is enabled. The wrapper returns a small sentinel tensor only while the fused selected-logprob path explicitly enables the skip flag. This avoids materializing the full `[B, S, V]` logits tensor when downstream code only needs selected next-token logprobs.

A forward pre-hook captures the hidden states entering the LM head, and the original `lm_head.forward` is retained and called whenever the skip flag is not active. The skip flag is reset in a `finally` block after model forward, so normal LM-head behavior is preserved outside the fused selected-logprob path.

Alternatives considered:

- Pre-hook alone: captures hidden states but still executes `lm_head.forward` and allocates `[B, S, V]`.
- Changing model outputs: would require broader model-family-specific changes.

The feature is opt-in only, local to the model instance, and covered by unit tests for a plain `nn.Linear` LM head.

# Limitations

This PR intentionally rejects unsupported combinations instead of silently falling back:

- context parallelism
- sequence packing
- top-k/top-p training-time filtering

End-to-end tests cover plain `nn.Linear` LM heads. LoRA-wrapped LM heads, tied embeddings, and DTensor-sharded LM-head weights are not covered by integration tests in this PR. The implementation keeps normal LM-head forwarding unchanged when the skip flag is inactive and handles DTensor weights through `.to_local()`, but I am happy to add more integration coverage if maintainers request it.

# Testing

```bash
ruff check --fix docs/guides/dpo.md docs/guides/grpo.md docs/guides/sft.md nemo_rl/models/policy/__init__.py nemo_rl/models/automodel/setup.py nemo_rl/models/automodel/train.py tests/unit/models/automodel/test_automodel_setup.py tests/unit/models/automodel/test_automodel_train.py
ruff format docs/guides/dpo.md docs/guides/grpo.md docs/guides/sft.md nemo_rl/models/policy/__init__.py nemo_rl/models/automodel/setup.py nemo_rl/models/automodel/train.py tests/unit/models/automodel/test_automodel_setup.py tests/unit/models/automodel/test_automodel_train.py
python -m pytest tests/unit/models/automodel/test_automodel_setup.py tests/unit/models/automodel/test_automodel_train.py -k "fused_linear_logprobs or precomputed_selected_logprobs" -v
```

# Additional Information

This follows the existing Megatron fused selected-token logprob concept, but uses AutoModel-specific hidden-state capture because AutoModel does not currently expose a backend-neutral API for skipping LM-head logits while returning the LM-head input.

### CODE REVIEW

File: docs/guides/grpo.md [hunk 1]
```diff
@@ -382,22 +382,35 @@ This works for GRPO because [ClippedPGLossFn](../../nemo_rl/algorithms/loss/loss
...
 ```
 
-**Notes:**
+For AutoModel, set the equivalent top-level policy keys:
+
+```yaml
+policy:
+  dtensor_cfg:
+    enabled: true
+  use_fused_linear_logprobs: true
```
Comment (yuki-97) [inline docs/guides/grpo.md:401]: not a blocker of this PR, but since after this PR both dtensor and mcore backend will support this, it makes more sense to move the two params to just under `policy`.

just curious @mohamed-alansary @pengdurice do you have interest to help do this after this PR?

**Review by yuki-97 — COMMENTED** [review]
@RayenTian to review, also cc @pengdurice if you have interest

File: docs/guides/grpo.md (hunk 1 above)
Comment (pengdurice) [reply docs/guides/grpo.md:401]: @yuki-97 , sure, I am happy to take on this;)

File: nemo_rl/models/automodel/train.py [hunk 2]
```diff
@@ -188,6 +189,179 @@ def apply_temperature_scaling(
...
+    if lm_head_path is None:
+        return
+    try:
+        lm_head = _get_module_by_path(model, lm_head_path)
+    except AttributeError:
+        return
+    setattr(lm_head, "_skip_lm_head_for_fused_linear_logprobs", enabled)
+
+
+def _tp_rank_for_fused_linear_logprobs(tp_group: Any) -> int:
```
Comment (RayenTian) [inline nemo_rl/models/automodel/train.py:240]: Could we move this topology and LM-head layout resolution out of the training path? This metadata is static and is already available in [setup.py](https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/models/automodel/setup.py) after AutoModel has applied TP/FSDP parallelization. 

What about defining a typed `FusedLinearLogprobsContext` in [automodel/config.py](https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/models/automodel/config.py), which could be populated during setup with the resolved LM-head path, chunk size, TP group/rank/size, and validated vocab-sharding layout. The training path could then consume that context and only retrieve the current LM-head weight, allowing these topology/layout helpers and the scattered _fused_linear_logprobs_* attributes to be removed. 

This would also make an unsharded LM head under TP fail during setup instead of being rediscovered—or misclassified—during every forward pass.

File: nemo_rl/models/policy/__init__.py [hunk 3]
```diff
@@ -484,6 +435,16 @@ class PolicyConfig(TypedDict):
     # If None, chunking is disabled and the full sequence is processed at once.
     logprob_chunk_size: NotRequired[int | None]
     generation: NotRequired[GenerationConfig]
+    # AutoModel-only opt-in for computing selected next-token logprobs directly
+    # from hidden states and the LM-head weight, avoiding full [B, S, V] logits
+    # for loss/logprob postprocessing. This mirrors the Megatron
+    # use_fused_linear_logprobs path. It is not compatible with context
+    # parallelism, sequence packing, or top-k/top-p training-time filtering.
+    use_fused_linear_logprobs: NotRequired[bool]
```
Comment (RayenTian) [inline nemo_rl/models/policy/__init__.py:443]: These new legacy TypedDict fields need defaults in the relevant exemplar YAMLs and matching [tests/unit/reference_configs](https://github.com/NVIDIA-NeMo/RL/tree/main/tests/unit/reference_configs) files.

File: nemo_rl/models/automodel/setup.py [hunk 4]
```diff
@@ -573,6 +600,98 @@ def _noop(*args, **kwargs) -> None:  # pragma: no cover
     _model_init._restore_loaded_model_dtype = _noop
 
 
+def _get_module_by_path(model: torch.nn.Module, module_path: str) -> torch.nn.Module:
```
Comment (RayenTian) [inline nemo_rl/models/automodel/setup.py:603]: Could we drop _get_module_by_path (defined identically here and in [train.py:192](https://github.com/NVIDIA-NeMo/RL/pull/3128/changes#diff-b193ca47f5466d66a35db24b0d7f1c96dca88663e65e938bba8837cec310d758R192)) and use PyTorch's built-in `nn.Module.get_submodule` instead?

File: nemo_rl/models/automodel/train.py [hunk 5]
```diff
@@ -188,6 +189,179 @@ def apply_temperature_scaling(
...
+        raise RuntimeError(
+            "policy.use_fused_linear_logprobs=True requires the LM-head weight "
+            "to be vocab-sharded on the tp mesh dimension when tensor "
+            "parallelism is enabled."
+        )
+
+    tp_rank = _tp_rank_for_fused_linear_logprobs(tp_group)
+    local_vocab_size = int(output_weight.shape[0])
+    vocab_start_index = tp_rank * local_vocab_size if local_vocab_is_tp_partition else 0
+    tensor_parallel_hidden_states = hidden_states.transpose(0, 1).contiguous()
```
Comment (RayenTian) [inline nemo_rl/models/automodel/train.py:350]: The post-parallelization LM-head pre-hook captures a Replicate() DTensor containing the full [B, S, H] activation on every TP rank. from_parallel_hidden_states_to_logprobs then all-gathers these already replicated activations and concatenates them along the sequence dimension, producing [TP*S, B, H]. Inference remains numerically correct but performs redundant communication and computation, while training backward fails due to the resulting sequence-dimension mismatch. 
Please use a fused implementation that accepts replicated hidden states, or reject tp_size > 1 until it is supported, and add a distributed forward/backward parity test.

File: nemo_rl/models/automodel/train.py [hunk 6]
```diff
@@ -348,18 +522,39 @@ def forward_with_post_processing_fn(
...
+            allow_flash_attn_args=allow_flash_attn_args,
+            use_cache=use_cache,
+        )
+    finally:
+        if use_fused_linear_logprobs:
+            _set_lm_head_skip_for_fused_linear_logprobs(model, False)
 
     # Extract logits from model outputs
-    logits = extract_logits(model, outputs)
+    logits = (
```
Comment (RayenTian) [inline nemo_rl/models/automodel/train.py:551]: The fused branch returns precomputed selected-token logprobs, not logits, but it is passed through the logits argument and later detected via ndim == 2. Could we make this contract explicit, for example by passing selected_logprobs separately or wrapping it in a small SelectedLogprobs type? This avoids relying on tensor rank to distinguish semantically different inputs.

File: nemo_rl/models/automodel/train.py [hunk 7]
```diff
@@ -348,18 +522,39 @@ def forward_with_post_processing_fn(
...
 
     # Model forward pass
-    outputs = model_forward(
-        model,
-        processed_inputs,
-        is_reward_model=is_reward_model,
-        allow_flash_attn_args=allow_flash_attn_args,
-        use_cache=use_cache,
-    )
+    if use_fused_linear_logprobs:
```
Comment (RayenTian) [inline nemo_rl/models/automodel/train.py:535]: Could we move this forward-scoped state management into `model_forward` and pass the fused-logprobs context explicitly? 

Clearing the captured hidden states, enabling the LM-head skip, and restoring it in finally all directly control the model(**model_args) invocation. Keeping that lifecycle next to the actual model call—ideally behind a context manager that restores the previous state—would simplify this orchestration layer and avoid exposing LM-head execution details here.

File: nemo_rl/models/automodel/train.py [hunk 8]
```diff
@@ -376,18 +571,30 @@ def forward_with_post_processing_fn(
...
     if isinstance(post_processing_fn, LossPostProcessor):
-        result, metrics = post_processing_fn(
-            logits=logits,
-            data_dict=data_dict,
-            processed_inputs=processed_inputs,
-            global_valid_seqs=global_valid_seqs,
-            global_valid_toks=global_valid_toks,
-            sequence_dim=sequence_dim,
-        )
+        loss_fn = post_processing_fn.loss_fn
```
Comment (RayenTian) [inline nemo_rl/models/automodel/train.py:579]: Could we make the precomputed-logprob input explicit instead of temporarily mutating `loss_fn.use_fused_linear_logprobs`? For example, define a small `SelectedLogprobs` wrapper in [algorithms/loss/interfaces.py](https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/algorithms/loss/interfaces.py), pass it through LossPostProcessor, and let [prepare_loss_input](https://github.com/NVIDIA-NeMo/RL/blob/4806fc0ae43c2b44505f90455e705d3ad2454874/nemo_rl/algorithms/loss/utils.py#L37). That keeps the logits-vs-logprobs distinction at the central input-conversion boundary and allows this save/set/restore block, the loss-function side-channel state, and shape-based input inference to be removed.

File: tests/unit/models/automodel/test_automodel_setup.py [hunk 9]
```diff
@@ -76,6 +139,60 @@ def mock_config():
     }
 
 
+@pytest.mark.automodel
+def test_setup_fused_linear_logprobs_does_not_duplicate_lm_head_parameters():
```
Comment (RayenTian) [inline tests/unit/models/automodel/test_automodel_setup.py:143]: This assertion does not detect duplicate LM-head registration because named_parameters() defaults to remove_duplicate=True. If the same lm_head were assigned to another module attribute, the returned names would remain unchanged and this test would still pass. Could we compare `named_parameters(remove_duplicate=False)` before and after, inspect `state_dict()` keys, or assert the registered module paths directly so the test actually enforces its stated invariant?

**Comment by RayenTian** [comment]
Thanks for adding this @mohamed-alansary ! Left some comments.
